### PR TITLE
Throw an exception without setting configuredProvider to null when ge…

### PR DESCRIPTION
**Pull request by benjamin-confino**
…tCDI returns null

Signed-off-by: Benjamin Confino <benjamic@uk.ibm.com>

### CODE REVIEW

**Comment by benjamin-confino** [comment]
This resolves #461 see also the discussion in #445

**Review by Emily-Jiang — APPROVED** [review]
LGTM!

File: api/src/main/java/jakarta/enterprise/inject/spi/CDI.java [hunk 1]
```diff
@@ -44,6 +44,7 @@
 public abstract class CDI<T> implements Instance<T> {
 
     private static final Object lock = new Object();
+    private static boolean providerSetManually = false;
```
Comment (manovotn) [inline api/src/main/java/jakarta/enterprise/inject/spi/CDI.java:47]: I think this should be `volatile`.

File: api/src/main/java/jakarta/enterprise/inject/spi/CDI.java [hunk 2]
```diff
@@ -79,6 +80,10 @@ private static CDIProvider getCDIProvider() {
             }
         } catch (IllegalStateException ignored) {
         }
+        //configuredProvider cannot be null if providerSetManually is true. If it somehow is an NPE is appropriate.
```
Comment (manovotn) [inline api/src/main/java/jakarta/enterprise/inject/spi/CDI.java:83]:
```suggestion
        //configuredProvider cannot be null if providerSetManually is true. If it somehow is, an ISE is appropriate.
```

File: api/src/main/java/jakarta/enterprise/inject/spi/CDI.java [hunk 3]
```diff
@@ -79,6 +80,10 @@ private static CDIProvider getCDIProvider() {
...
+        if (providerSetManually && configuredProvider.getCDI() == null) {
```
Comment (manovotn) [inline api/src/main/java/jakarta/enterprise/inject/spi/CDI.java:84]: You could add this logic directly into the `catch` block and just check for `providetSetManually`.

**Review by manovotn — CHANGES_REQUESTED** [review]

File: api/src/main/java/jakarta/enterprise/inject/spi/CDI.java (hunk 1 above)
Comment (benjamin-confino) [reply api/src/main/java/jakarta/enterprise/inject/spi/CDI.java]: I've made the change.

File: api/src/main/java/jakarta/enterprise/inject/spi/CDI.java (hunk 3 above)
Comment (benjamin-confino) [reply api/src/main/java/jakarta/enterprise/inject/spi/CDI.java]: Good idea!

**Comment by benjamin-confino** [comment]
I have uploaded changes based on suggestions from @manovotn

**Review by manovotn — APPROVED** [review]

**Comment by benjamin-confino** [comment]
Hello. There has not been any movement on this for a while. Is anything else needed or can this be merged?

**Pull request merged**